pennylane/transforms/diagonalize_measurements.py

```diff
@@ -0,0 +1,345 @@
+"""Transform to diagonalize measurements on a tape, assuming all measurements are commuting."""
+
+from copy import copy
+from functools import singledispatch
+
+import pennylane as qml
+from pennylane.operation import Tensor
+from pennylane.ops import CompositeOp, LinearCombination, SymbolicOp
+from pennylane.transforms.core import transform
+
+# pylint: disable=protected-access
+
+_default_supported_obs = frozenset(["PauliZ", "Identity"])
+
+
+def null_postprocessing(results):
+    """A postprocessing function returned by a transform that only converts the batch of results
+    into a result for a single ``QuantumTape``.
+    """
+    return results[0]
+
+
+@transform
+def diagonalize_measurements(tape, supported_base_obs=("PauliZ", "Identity")):
+    """Diagonalize the measurements on a tape if they are not supported. Raises an error if the
+    measurements do not commute.
```

Comment:
I think I need to understand the commuting constraint a bit more. It makes sense to me if we have two observables `A` and `B` that they must commute for us to jointly diagonalize them. But it looks like there are constraints on `A` and `B` individually. For example `A = Z(0) + X(0)` raises an error:
```python
measurements = [
    qml.expval(qml.Z(0) + qml.X(0))
]
tape = qml.tape.QuantumScript(measurements=measurements)
tapes, processing_fn = diagonalize_measurements(tape)
```
```
ValueError: Expected measurements on the same wire to commute, but X(0) overlaps with another non-commuting observable on the tape.
```

Reply:
This transform should never get `Z(0) + X(0)`, because it assumes `split_non_commuting` has been applied, and those will have been put on separate tapes. 

It's true that `X(0) + Z(0)` has a diagonalizing gate, but 

1. if we want to use `observable` instead of `eigvals` (so that we can send the original measurement type to the backend instead of converting to samples, then I would expect it won't work, because the diagonalizing gate of `X(0) + Z(0)` doesn't diagonalize to `Z(0) + Z(0)`, it diagonalizes to a single sample readout on wire 0 with `eigvals` applied in post-processing. 
2. if we want to be able to selectively diagonalize only some of {X, Y, Z, Hadamard}, then we need to diagonalize terms individually, and once we apply `X(0).diagonalzing_gates()`, `Z(0)` won't work.

So based on how this transform works, `X(0) + Z(0)` can't be diagonalized without first being split into separate tapes.

Reply:
Thanks! I guess this is all fine, but where I'm struggling is with the name of the function or that it sounds more general than it is.

So this function does:
- Input is a circuit which only has `X`, `Y`, `Z`, or `H` measurements on each wire
- Output is a circuit which only has observables specified by `supported_base_obs` on each wire. The circuit has rotations to account for the observables not in `supported_base_obs`.

Is that a fair summary?

Reply:
Yes, that sounds right. If you take a tape and put it through `split_non_commuting` and then this transform, and tell it that only `X` and `Z` are supported, it will apply diagonalizing gates for any `Y` or `Hadamard` observables and then swap them with `Z`.

Maybe as suggested in the meeting, `diagonalize_unsupported_observables` would be a better name?

Reply:
decided offline on doc string clarification only